Source/Scene/ModelExperimental/CustomShader.js

```diff
@@ -0,0 +1,298 @@
+ */
+export default function CustomShader(options) {
+  options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+  this.mode = defaultValue(options.mode, CustomShaderMode.MODIFY_MATERIAL);
+  this.lightingModel = options.lightingModel;
+  this.uniforms = defaultValue(options.uniforms, defaultValue.EMPTY_OBJECT);
+  this.varyings = defaultValue(options.varyings, defaultValue.EMPTY_OBJECT);
+  this.vertexShaderText = options.vertexShaderText;
+  this.fragmentShaderText = options.fragmentShaderText;
+
+  this._defaultTexture = undefined;
+  this._textures = {};
+  this._loadedImages = [];
+
+  this.uniformMap = buildUniformMap(this);
+
+  // Lists of variables used from the automatically-generated structs. These
+  // can be used for optimizations when generating the overall shader.
+  // The values are JS objects used like sets. The keys are the variable names,
+  // the values do not matter here. This is for three reasons:
+  // 1. We are not using ES6 sets
+  // 2. Using a dictionary automatically de-duplicates variable names
+  // 3. It makes it easy to do queries such as:
+  //    if ("position" in attributeSet) { ... }
+  this._usedVariablesVertex = {
+    // attributes from the glTF
+    attributeSet: {},
+  };
+  this._usedVariablesFragment = {
+    // positions in various reference frames, e.g. positionMC, positionEC
+    positionSet: {},
+    // attributes from the glTF
+    attributeSet: {},
+    // properties of the material used
+    materialSet: {},
+  };
+  findUsedVariables(this);
+}
+
+function buildUniformMap(customShader) {
+  var uniforms = customShader.uniforms;
+  var uniformMap = {};
+  for (var uniformName in uniforms) {
+    if (uniforms.hasOwnProperty(uniformName)) {
+      var uniform = uniforms[uniformName];
+      var type = uniform.type;
+      //>>includeStart('debug', pragmas.debug);
+      if (type === UniformType.SAMPLER_CUBE) {
+        throw new DeveloperError(
+          "CustomShader does not support samplerCube uniforms"
+        );
+      }
+      //>>includeEnd('debug');
+
+      if (type === UniformType.SAMPLER_2D) {
+        fetchTexture2D(customShader, uniformName, uniform.value);
+        uniformMap[uniformName] = createUniformTexture2DFunction(
+          customShader,
+          uniformName
+        );
+      } else {
+        uniformMap[uniformName] = createUniformFunction(
+          customShader,
+          uniformName
+        );
+      }
+    }
+  }
+  return uniformMap;
+}
+
+function createUniformTexture2DFunction(customShader, uniformName) {
+  return function () {
+    return customShader._textures[uniformName];
+  };
+}
+
+function createUniformFunction(customShader, uniformName) {
+  return function () {
+    return customShader.uniforms[uniformName].value;
+  };
+}
+
+function fetchTexture2D(customShader, uniformName, textureValue) {
```

Comment:
This texture fetching code is a simplification of the rather convoluted code from `Material.js`.

@lilleyse I feel like we should have a more general image loader (GltfImageLoader is too gltf-specific) that can handle loading from a URI string, Resource, and other things (videos, KTX2 like the old Material, array buffers). What do you think? Though this is beyond the scope of this PR

Reply:
https://github.com/CesiumGS/cesium/issues/9741 -- opened this issue for discussion

Reply:
We talked about this on Friday. For now, I'm going to at least isolate this code into a `TextureManager` class. The caller is responsible for calling destroy() on the CustomShader (because it's hard to tell the intent if a custom shader is used for multiple tilesets)